File: p23/p23.py
# http://projecteuler.net/problem=23

def get_divisors(n):
    """ Return the divisors of n that are less than n """
    dvs = []
    for i in range(1, int(n ** 0.5) + 1):
        if n % i == 0:
            dvs.append(i)
            if n / i != i:
                dvs.append(n / i)
    dvs.remove(n)
    return dvs

def is_abundant(n):
    """ Returns true if n is abundant.
        An abundant number is a number whose factors, when summed up, is less
        than this sum. """
    return n < sum(get_divisors(n)) 


# Get all abundant numbers below 28,124
total = 0
abundant_nums = set(i for i in range(12, 28124) if is_abundant(i))

# Find all numbers below 28,124 that can't be represented as a sum
# of two abundant numbers
for i in range(1, 28124):
    found = False
    for abundant_num in abundant_nums:
        diff = i - abundant_num
        if diff in abundant_nums:
            found = True
            break

    if not found:
        total += i

print total

File: csharp/P21_to_30/P23.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Numerics;
using System.Text;
using System.Threading.Tasks;
using ProjectEuler.Helpers;

namespace ProjectEuler.P21_to_30
{
    /// <summary>
    /// https://projecteuler.net/problem=23
    /// </summary
    internal class P23
    {
        private const int Max = 28123;

        /// <summary>
        /// Calculates the sum of all positive integers that cannot be written as the sum of
        /// two abundant numbers.
        /// </summary>
        public int Solve()
        {
            var abundantNumbers = Enumerable.Range(1, Max).Where(IsAbundant).ToArray();
            var set = new HashSet<int>(abundantNumbers);
            var sum = 0;

            for (var i = 0; i <= Max; i++)
            {
                if (!abundantNumbers.Any(x => set.Contains(i - x)))
                {
                    // i cannot be expressed as sum of two abundant numbers; add it to the total
                    sum += i;
                }
            }

            return sum;
        }

        /// <summary>
        /// Returns true if the sum of the proper divisors of n is greater than n.
        /// </summary>
        private bool IsAbundant(int n)
        {
            return PrimeHelpers.Divisors(n).Sum() - n > n;
        }
    }
}